File: 0001-0500/0037/0037.sudoku-solver.cs
/*
 * @lc app=leetcode id=37 lang=csharp
 *
 * [37] Sudoku Solver
 */

// @lc code=start
public class Solution {
    public void SolveSudoku(char[][] board) {
        int[][] rows = new int[9][];
        int[][] cols = new int[9][];
        int[][] boxes = new int[9][];
        for (int i = 0; i < 9; i++) {
            rows[i] = new int[9];
            cols[i] = new int[9];
            boxes[i] = new int[9];
        }
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j ++) {
                if (board[i][j] != '.') {
                    int num = board[i][j] - '1';
                    rows[i][num] = 1;
                    cols[j][num] = 1;
                    boxes[i / 3 * 3 + j / 3][num] = 1;
                }
            }
        }
        Solve(board, rows, cols, boxes, 0, 0);
    }

    private bool Solve(char[][] board, int[][] rows, int[][] cols, int[][] boxes, int i, int j) {
        if (i == 9) {
            return true;
        }
        if (j == 9) {
            return Solve(board, rows, cols, boxes, i + 1, 0);
        }
        if (board[i][j] != '.') {
            return Solve(board, rows, cols, boxes, i, j + 1);
        }
        for (int num = 0; num < 9; num++) {
            if (rows[i][num] == 1 || cols[j][num] == 1 || boxes[i / 3 * 3 + j / 3][num] == 1) {
                continue;
            }
            board[i][j] = (char)(num + '1');
            rows[i][num] = 1;
            cols[j][num] = 1;
            boxes[i / 3 * 3 + j / 3][num] = 1;
            if (Solve(board, rows, cols, boxes, i, j + 1)) {
                return true;
            }
            board[i][j] = '.';
            rows[i][num] = 0;
            cols[j][num] = 0;
            boxes[i / 3 * 3 + j / 3][num] = 0;
        }
        return false;
    }
}
// @lc code=end

File: 0001-0500/0037/0037.sudoku-solver.cpp
/*
 * @lc app=leetcode id=37 lang=cpp
 *
 * [37] Sudoku Solver
 */

// @lc code=start
class Solution
{
public:
    void solveSudoku(vector<vector<char>> &board)
    {
        vector<vector<bool>> row(9, vector<bool>(9, false));
        vector<vector<bool>> col(9, vector<bool>(9, false));
        vector<vector<bool>> box(9, vector<bool>(9, false));
        vector<pair<int, int>> empty;
        for (int i = 0; i < 9; ++i)
        {
            for (int j = 0; j < 9; ++j)
            {
                if (board[i][j] == '.')
                {
                    empty.push_back({i, j});
                }
                else
                {
                    int num = board[i][j] - '1';
                    row[i][num] = true;
                    col[j][num] = true;
                    box[i / 3 * 3 + j / 3][num] = true;
                }
            }
        }
        dfs(board, empty, row, col, box, 0);
    }

    bool dfs(vector<vector<char>> &board, vector<pair<int, int>> &empty, vector<vector<bool>> &row, vector<vector<bool>> &col, vector<vector<bool>> &box, int index)
    {
        if (index == empty.size())
        {
            return true;
        }
        int i = empty[index].first;
        int j = empty[index].second;
        for (int num = 0; num < 9; ++num)
        {
            if (!row[i][num] && !col[j][num] && !box[i / 3 * 3 + j / 3][num])
            {
                board[i][j] = num + '1';
                row[i][num] = true;
                col[j][num] = true;
                box[i / 3 * 3 + j / 3][num] = true;
                if (dfs(board, empty, row, col, box, index + 1))
                {
                    return true;
                }
                board[i][j] = '.';
                row[i][num] = false;
                col[j][num] = false;
                box[i / 3 * 3 + j / 3][num] = false;
            }
        }
        return false;
    }
};
// @lc code=end
